zcash_client_backend/src/zip321.rs

```diff
@@ -0,0 +1,1046 @@
+
+        tuple((
+            paramname,
+            opt(preceded(
+                char('.'),
+                recognize(tuple((
+                    one_of("123456789"),
+                    map_opt(digit0, |s: &str| if s.len() > 3 { None } else { Some(s) }),
+                ))),
+            )),
+        ))(input)
+    }
+
+    pub fn parse_amount<'a>(input: &'a str) -> IResult<&'a str, Amount> {
+        map_res(
+            tuple((
+                digit1,
+                opt(preceded(
+                    char('.'),
+                    map_opt(digit0, |s: &str| if s.len() > 8 { None } else { Some(s) }),
+                )),
+            )),
+            |(whole_s, decimal_s): (&str, Option<&str>)| {
+                let coins: i64 = whole_s
+                    .to_string()
+                    .parse::<i64>()
+                    .map_err(|e| e.to_string())?;
+
+                let zats: i64 = match decimal_s {
+                    Some(d) => format!("{:0<8}", d)
+                        .parse::<i64>()
+                        .map_err(|e| e.to_string())?,
+                    None => 0,
+                };
+
+                if coins >= 21000000 && (coins > 21000000 || zats > 0) {
+                    return Err(format!(
+                        "{} coins exceeds the maximum possible Zcash value.",
+                        coins
+                    ));
+                }
+
+                let amt = coins * COIN + zats;
+
+                Amount::from_nonnegative_i64(amt)
+                    .map_err(|_| format!("Not a valid zat amount: {}", amt))
+            },
+        )(input)
+    }
+
+    fn to_indexed_param<'a, P: consensus::Parameters>(
+        params: &'a P,
+        ((name, iopt), value): ((&str, Option<&str>), &str),
+    ) -> Result<IndexedParam, String> {
+        let param = match name {
+            "address" => Address::parse(params, value)
+                .map_err(|e| match e {
+                    AddrParseError::TAddrError(_) => format!(
+                        "Could not interpret {} as a valid transparent address.",
+                        value
+                    ),
+                    AddrParseError::SAddrError(_) => {
+                        format!("Could not interpret {} as a valid Sapling address.", value)
+                    }
+                })
+                .and_then(|addr_opt| {
+                    addr_opt.map(Param::Addr).ok_or(format!(
+                        "Could not interpret {} as a valid Zcash address.",
+                        value
+                    ))
+                }),
+
+            "amount" => parse_amount(value)
+                .map(|(_, a)| Param::Amount(a))
+                .map_err(|e| e.to_string()),
+
+            "label" => percent_decode(value.as_bytes())
+                .decode_utf8()
+                .map(|s| Param::Label(s.into_owned()))
+                .map_err(|e| e.to_string()),
+
+            "message" => percent_decode(value.as_bytes())
```

Comment:
I dislike `message` as a parameter name, because it seems very likely to be confused with `memo`: users looking at a URI may think that `message` _is_ the memo. This could maybe be rectified by not encoding `message` in a human-readable format (e.g. also base64 it), but I think that defeats the intended purpose of `message`, so renaming the parameter makes more sense to me.

`label` is similarly non-descriptive and prone to confusion with `message`. I note that AFAICT the spec requires a `message` to be associated with a `recipient`, despite describing it as "descriptive information about the transaction". It feels like the overall intent is for `label` to be descriptions of recipients, and `message` is for a decription of the whole transaction, so maybe they should be named to indicate this?

Reply:
This is actually taken from BIP-21. I will clarify the spec with respect to each of these though.

Reply:
Hmm, okay. I still don't like it, but If we're explicitly wanting BIP-21 compatibility then I guess this is fine.